ToDo REST API 만들기 과제 제출

csh9066 opened the pull request:
리뷰 요청 부탁드립니다!
평소 풀리퀘스트를 보낸적이 없어서 제대로 보낸건지 잘 모르겠네요 ㅜ

csh9066 committed:
ToDo REST API 만들기 과제 제출

csh9066 commented:
핸들러에 toJson 이라는 object에서 string을 변환하는 메소드가 있는데 이름을 잘못 지은거 같아요
toString 으로 바꿀거 같아요 풀리퀘스트 수정하는 법을 몰라서 ㅜㅜ

hannut91 commented:
잘하셨습니다 ㅎㅎ 이렇게 하는 것 맞습니다.

풀 리퀘스트는 따로 수정할 필요 없습니다 ㅎㅎ 지금 작업중인 브랜치에서 새로운 커밋을 하고 푸시하면, 자동으로 반영이 됩니다!

hannut91 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,33 @@
+        System.out.println(new App().getGreeting());
+
+        try {
+            InetSocketAddress address = new InetSocketAddress(8000);
```
8000은 `포트번호`를 의미하고 있는 것 같아요. 8000이 단순히 숫자가 아니라 특별한 의미를 갖는다는 것을 상수로 표현해보는 것은 어떨까요?

See also

- [Replace Magic Literal](https://refactoring.com/catalog/replaceMagicLiteral.html)

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,142 @@
+import java.util.stream.Collectors;
+
+public class DemoHttpHandler implements HttpHandler {
+    private TaskRepository taskRepository = new TaskRepository();
+    ObjectMapper objectMapper = new ObjectMapper();
+
+    private final String TASK_DEFAULT_PATH = "/tasks";
```
private으로 선언한 것과, public(기본으로 선언된) 것들이 섞여 있는 것 같아요. 구분해주면 어떨까요?

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,142 @@
+     * path가 올바른 pathvarialbe을 포함하고 있는지 확인하는 메소드
+     * 좀 더 좋은 좋은 이름을 지을 수 있을거 같은데 생각이 안남 ㅜㅜ
+     */
+    private boolean containPathVariable(String path) {
```
좋은 메서드 이름을 고민해 주셨네요 ㅎㅎ 지금 당장에 아주 좋은 이름을 짓는건 사실 굉장히 어려운 일이이에요. 왜냐하면 우리는 프로젝트를 점점 진행하면서 배우게 되기 때문이에요. 지금 엄청 좋은 이름을 생각해냈다고 하더라도, 나중에 새로 알게된 사실 때문에 변수 이름이 이상해지는 경우도 있죠.

제가 생각하기에는 지금도 나쁘지 않다고 생각해요 ㅎㅎ 과제를 진행하면서, 배운다음에 더 좋은 이름이 생겼을 때 다시 더 좋은 이름으로 바꾸는 것도 좋다고 생각해요

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,142 @@
+     */
+    private boolean containPathVariable(String path) {
+
+        String substring = path.substring(TASK_DEFAULT_PATH.length());// ex) /tasks/3 -> /3
```
코드로 충분히 설명이 어려워서 친절하게 주석을 남겨주셨네요! 나중에 이 부분을 테스트로 옮긴다면 어떨까? 고민해 보시면 좋을 것 같아요. 테스트는 아직 배우지 않지만요!

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,142 @@
+            return false;
+        }
+
+        if (substring.charAt(0) != '/') {
```
좀 더 의도를 드러내 본다면, `/`으로 `시작하는지 아닌지` 인 것 같아요. 그래서 

```
substring.startWith("/")
```

이런 모습이면 더 좋을 것 같은데 이런게 있을까요?

csh9066 replied:
평소 사용하던 것만 사용해서 startWith, endWith가 있는 걸 까먹고 있었네요 
요번 기회에 한번 스트링 클래스 메소드에 대해 정리좀 해야겠어요!

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,142 @@
+            return false;
+        }
+        
+        String pathVariable = substring.substring(1);
```
왜 1인지 이해하기 어려울 것 같네요. 그렇다고 상수로 분리하라는 것은 아닙니다

csh9066 replied:
한줄이지만 이해하기 쉬운 이름으로 메소드로 분리해도 상관이 없을까요?

hannut91 replied:
네 ㅎㅎ 시도해 보세요

hannut91 commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -0,0 +1,41 @@
+
+    public Long save(Task task) {
+        if (task.getId() == null) {
+            task.setId(++sequence);
```
전위연산자를 사용해서 값이 변경되기 전의 값을 할당하고 나서 값을 증가시키고 있는 것 같아요. 사용법을 정확히 알고 있다면 ++value 와 value++ 을 구분해서 사용할 수 있겠지만, 사실 조금 헷갈릴수도 있는 문법이에요. 우리가 한 번 더 생각해야되는거죠. 아주 조금이라도 헷갈릴 수 있거나 다르게 해석될 여지가 있다면 저는 사용을 지양하는 것이 좋다고 생각해요.

csh9066 replied:
제가 생각해도 읽기 힘들고 헷갈릴 수 있을거 같아요!
좋은 변수명을 선언해서 읽기 좋게 만들어 볼게요

좋은 리뷰 감사합니다! 앞으로도 잘 부탁드립니다 :)

hannut91 replied:
설명도 틀렸는데 눈치를 못채셨군요 ㅎㅎ 전위연산자는 값을 변경하고나서 할당하죠

csh9066 committed:
refactor: App에 매직넘버를 상수로 선언

서버의 포트번호가 매직넘버에서 상수로 추출했습니다

csh9066 committed:
refactor: DemoHttpHandler의 toJson 메소드 이름 변경

기존에 있던 toJson라고 이름을 잘못 정했는다 String을
반환하는 메소드기 떄문에 toString으로 이름을 변경했습니다.

csh9066 committed:
refactor: DemoHttpHandler의 필드변수 objectMapper 접근 지정자 추가

objectMapper가 접근 지정자가 default로 되어있었는데 현재 클래스에서만 사용하기 때문에 private로 변경 했다.

hannut91 commented:
이번주도 파이팅입니다 ㅎㅎ 도움이 필요하시면 말씀해 주세요

csh9066 committed:
refactor: TaskRepository의 save 변경

save 시 sequence를 전위연산자를 이용해 증가시켰는데 좀 더 읽기 쉽게 메소드를 만듬

csh9066 committed:
feat: HttpStatus enum 추가

csh9066 committed:
feat: DemoHttpHandler의 task가 존재하지 않을 떄 요청 처리 추가

csh9066 committed:
refator: DemoHttpHandler의 containPathVariable 메소드 변경

csh9066 commented:
리뷰 요청 부탁드립니다. 첫 리뷰 요청 후 이런걸 해봤습니다!

1. 윤석님께 말씀해주신 피드백 바탕으로 리팩토링 해봤습니다.
2. 모든 테스트 통과하도록 기능 추가를 했습니다.
3. 좋은 커밋 메시지를 작성해볼려고 했습니다.

피드백을 제대로 반영했는지, 리뷰어님께서 보시기에 올바른 커밋메시지인지 궁금합니다!

csh9066 committed:
refator: DemoHttpHandler toString 메소드 이름 변경

다시 생각해보니 object를 string 형식의 json으로 바꾸니까 toJson이 맞는거 같다

hannut91 commented:
커밋 메시지는 제목에는 `이유, 의도`를, 내용에는 그 의도를 실현하기 위해서 실제로 한 일(`무엇`)을 자세하게 설명합니다.

```
refator: DemoHttpHandler toString 메소드 이름 변경

다시 생각해보니 object를 string 형식의 json으로 바꾸니까 toJson이 맞는거 같다
```

저라면 이렇게 적었을 것 같아요


```
JSON형식으로 변경하는 메서드의 이름을 더 명확하게 변경하라

지금은 객체를 JSON문자열로 변경하는 메서드의 이름이 toString()입니다.
하지만 사실 문자열이 중요한 것이 아니라, JSON문자열로 바뀌는 것이 더 중요한 사실입니다.
그래서 JSON문자열로 변경한다는 것을 의도를 드러내기 위해서 `toJson`으로 변경했습니다.
```

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,157 @@
+        // POST /tasks
+        if (method.equals("POST") && path.equals(TASK_DEFAULT_PATH)) {
+            Task task = toTask(body);
+            Long savedId = taskRepository.save(task);
+
+            Task savedTask = taskRepository.findById(savedId);
```
`save`를 했을 때 생성된 Task를 반환해도 좋을 것 같네요

csh9066 replied:
명령과 쿼리를 분리 해봤는데 별로 좋은 방법은 아닌가요?
아니면 제가 잘못 사용하고 있는건지 궁금합니다!

hannut91 replied:
음 단순히 명령과 조회를 구분한다기 보다, 모델을 구분하는게 더 중요한 것 같아요.

hannut91 commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -0,0 +1,44 @@
+    private Long sequence = 1L;
+    private Map<Long, Task> taskMap = new HashMap<>();
+
+    public Task findById(Long id) {
+        return taskMap.get(id);
+    }
```
`taskMap`에서 우리가 원하는 task를 찾지 못할 경우 `null`을 반환하고 있는 것 같아요.이 함수를 사용하는 입장에서는 이 함수가 `null`을 반환할 수 있는지 예상할 수 없고, 그리고 `null`처리를 하지 않아도 자바 컴파일러는 아무런 불만도 표현하지 않기 때문에 위험성이 있는 것 같아요. 만약 반환 타입을 Optional 그대로 반환한다면, 함수의 Signature에서 이 값이 있을 수도 있고 없을 수도 있다는 의도를 드러내고 있고, 없는 경우를 처리하지 않으면 컴파일도 되지 않아서 에러 처리를 강제할 수 있어요

하지만 Optional 반환은 다른 부분이 어떻게 만들어졌냐가 관건이라고 보는데, 함수형으로 처리 중이 아니라면 예외가 가장 일반적이죠. Optional을 리턴하면 사용하는 쪽에서 비슷한 코드를 작성해야 하기 때문에 사용하기 힘든 인터페이스가 될 가능성이 커요.

csh9066 replied:
`null`을 반환할 수 있는 메소드는 옵셔널으로 반환하는게  좋다라고 이해했습니다. 

> `Optional`을 리턴하면 사용하는 쪽에서 비슷한 코드를 작성해야 하기 때문에 사용하기 힘든 인터페이스가 될 가능성이 커요.

이 부분은 제가 이해를 못했는데 다시 한번 설명 해주실 수 있나요?

csh9066 replied:
`Optinal` 로 코드를 수정하면서 무슨 말씀인지 이해했습니다.
`Optional`로 반환하면 보통 orElseThrow로 예외를 발생시키는데
만약 예외를 발생시키지 않고 함수형으로 처리를 하지 않으면 조금 까다롭네요 ㅜㅜ

`Optional`를 사용하는 입장에서 함수형을 사용하지 않거나 예외를 반환하지 않고 싶을 때
이런 방법은 어떻게 생각하시나요?

```java
Task task = taskRepository.findById(id).get();
if (task == null) {
    null 일때 로직
}
```

과제는 예외처리를 던지는 쪽으로 처리 했습니다만 `Optional` 이 익숙하지 않아서 이런방법을 사용해도 되는지 모르겠네요

hannut91 replied:
`Optional`을 사용하지 않고, 예외를 반환하지 않을 때는 `에러 객체`를 만들어서 반환하는 방법이 있습니다만, 자바에서 보통 그렇게 하지는 않죠. 예외가 없는 golang에서는 애초에 에러를 같이 반환합니다.

```golang
task, err := findByTestId(id)
if err == nil {
  // Do something
}
```

> Optional을 리턴하면 사용하는 쪽에서 비슷한 코드를 작성해야 하기 때문에 사용하기 힘든 인터페이스가 될 가능성이 커요.

의 의미는 Optional을 사용하게 되면 이 것을 받는 쪽은 모두 Optional을 처리하도록 코드를 짜게 될거라는 겁니다. 즉 일관성이 깨질 수 있어요.(어디는 Optional, 어디는 try catch) 그래서 주의하라는 말씀이었습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public Long update(Long id, Task task) {
+        Task findTask = taskMap.get(id);
+        findTask.setTitle(task.getTitle());
```
여기서도 사실 null일 수 있는 가능성이 있는거죠. 만약 update를 하기전에 확인하지 않고 사용했다면, NPE가 발생했겠죠

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,157 @@
+        // DELETE /tasks/{id}
+        if (method.equals("DELETE") && containPathVariable(path)) {
+            Long id = getPathVariable(path);
+            Task findTask = taskRepository.findById(id);
```
`findTask`는 할 일을 찾아라! 라고하는 메서드 이름 같아보이네요. `task`라고만 사용해도 여기서는 이해할 수 있을 것 같은데 어떠세요?

csh9066 committed:
refator: TaskRepository save 메소드의 반환값을 변경해라

지금은 save 메소드 호출 시 Long 타입인 id를 반환 합니다. 좀 더 범용적으로 사용하기 위해 save 호출 시 TaskRepository에 저장된 Task를 반환하도록 변경했습니다.

csh9066 committed:
refactor: DemoHttpHandler에 필요없는 주석을 제거해라

csh9066 committed:
feat: TaskRepository findById 호출시 옵셔널을 반환해라

findById 호출 시 null을 반환할 수 있는데 기존 코드에는 Task를 반환하도록 했다. Optional을 반환해 의도가 명확하도록 했다.

csh9066 committed:
refactor: TaskRepository에 사용하지 않는 update를 제거해라

기존 Task update시 TaskRepository update를 했지만 Task 객체를 변경하는 방법으로 변경해서 사용하지 않는 메소드여서 삭제 했습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -0,0 +1,35 @@
+
+public class TaskRepository {
+    private Long sequence = 1L;
+    private Map<Long, Task> taskMap = new HashMap<>();
```
hashMap을 사용해서 특정 할 일을 찾을 때 엄청 빠르겠네요 ㅎㅎ 하지만 만약 사용자가 등록된 순서대로 할 일 목록을 받길 원한다면 어떻게 될까요?

csh9066 replied:
단순히 등록된 순서만 받길 원한다면 List 컬렉션을 이용할거 같아요!
만약 검색도 빠르게하고 싶고 순서대로 받고 싶으면 LinkedHashMap을 사용할거 같아요!

hannut91 reviewed (approved):
고생하셨습니다 ㅎㅎ 과제풀이 영상을 통해서 복습해 보세요. 문제를 어떻게 해결해 나가는지 과정을 눈여겨 보는게 중요합니다

csh9066 committed:
[refactor] PathVariable을 확인하는 메소드를 정규표현식으로 구현하여라

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,128 @@
+            // PUT /tasks/{id}
+            if (method.equals("PUT") && containPathVariable(path)) {
+                Long id = getPathVariable(path);
+                Task task = findTaskById(id);
+                Task taskToUpdate = toTask(body);
+                task.setTitle(taskToUpdate.getTitle());
```
```suggestion
                Task target = findTaskById(id);
                Task source = toTask(body);
                target.setTitle(source.getTitle());
```

target <-> source 을 사용해서 표현해보았는데 어떤 것 같으세요?

csh9066 replied:
정말 좋은 변수명 같아요! 이제부터 자주 사용할 변수명이 될거 같네요! 감사합니다 :)

csh9066 committed:
Update app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java

Co-authored-by: Yunseok <hannut91@gmail.com>
